### PR TITLE
fix: Allow nested `is_in()` in `when()/then()` for full-streaming

### PR DESCRIPTION
**Pull request by nameexhaustion**
Fixes https://github.com/pola-rs/polars/issues/15767

This PR also does some refactoring to consolidate and improve the way we identify expressions as elementwise.

#### Replacing `streamable` with `is_elementwise`

A lot of places where we currently use the `streamable` terminology actually needed to use an even stricter `elementwise` requirement (e.g. filter / slice pushdown. `streamable` expressions include all elementwise expressions and more. For example, `filter()` / `explode()` expressions are streamable, but they are not elementwise.

Where this matters is if the result of a streamable expression needs to be projected next to other result columns, which is almost always the case. The existing in-memory engine and evaluators cannot do this properly as it projects the result columns independently within every chunk, which could lead to height mismatches. The only place that properly supports this is the new-streaming engine, where the `ZipNode` performs the projection properly.

### CODE REVIEW

File: crates/polars-plan/src/plans/aexpr/mod.rs [hunk 1]
```diff
@@ -218,35 +219,41 @@ impl AExpr {
...
     }
-    /// Any expression that is sensitive to the number of elements in a group
-    /// - Aggregations
-    /// - Sorts
-    /// - Counts
-    /// - ..
-    pub(crate) fn groups_sensitive(&self) -> bool {
+
+    /// Checks whether this expression is elementwise. This only checks the top level expression.
+    pub(crate) fn is_elementwise_top_level(&self) -> bool {
```
Comment (nameexhaustion) [inline crates/polars-plan/src/plans/aexpr/mod.rs:224]: `is_elementwise_top_level()` replaces `groups_sensitive()`

File: crates/polars-plan/src/plans/aexpr/utils.rs [hunk 2]
```diff
@@ -1,121 +1,144 @@
...
 
-fn has_series_or_range(ae: &AExpr) -> bool {
-    matches!(
-        ae,
-        AExpr::Literal(LiteralValue::Series(_) | LiteralValue::Range { .. })
-    )
-}
+/// Checks if the top-level expression node is elementwise. If this is the case, then `stack` will
+/// be extended further with any nested expression nodes.
+pub fn is_elementwise(stack: &mut Vec<Node>, ae: &AExpr, expr_arena: &Arena<AExpr>) -> bool {
```
Comment (nameexhaustion) [inline crates/polars-plan/src/plans/aexpr/utils.rs:5]: Base `is_elementwise()` function - recursion is done through a nodes `stack` which allows the caller to inspect the exprs.

File: crates/polars-plan/src/plans/aexpr/utils.rs [hunk 3]
```diff
@@ -1,121 +1,144 @@
...
 
-impl Default for IsStreamableContext {
-    fn default() -> Self {
-        Self {
-            flags: StreamableFlags::all(),
-            context: Default::default(),
-        }
+                if matches!(expr_arena.get(rhs), AExpr::Literal { .. }) {
+                    stack.push_node(input.first().unwrap().node());
+                    return;
```
Comment (nameexhaustion) [inline crates/polars-plan/src/plans/aexpr/utils.rs:27]: This is the fix for the `when(A.is_in(B))` issue - we don't traverse into RHS literals of `is_in()`. We had this already in predicate pushdown but it has been moved here.

File: crates/polars-plan/src/plans/aexpr/mod.rs [hunk 4]
```diff
@@ -218,35 +219,41 @@ impl AExpr {
...
+            // Non-strict strptime must be done in-memory to ensure the format
+            // is consistent across the entire dataframe.
+            #[cfg(feature = "strings")]
+            Function {
+                options,
+                function: FunctionExpr::StringExpr(StringFunction::Strptime(_, opts)),
+                ..
+            } => {
+                assert!(options.is_elementwise());
+                opts.strict
```
Comment (nameexhaustion) [inline crates/polars-plan/src/plans/aexpr/mod.rs:239]: This rule was moved from new-streaming below

File: crates/polars-plan/src/plans/optimizer/predicate_pushdown/mod.rs [hunk 5]
```diff
@@ -632,34 +657,32 @@ impl<'a> PredicatePushDown<'a> {
...
-                acc_predicates.retain(|_, predicate| {
-                    if has_aexpr(predicate.node(), expr_arena, |ae| matches!(ae, AExpr::Len)) {
-                        local_predicates.push(predicate.clone());
-                        false
-                    } else {
-                        true
+                if cfg!(debug_assertions) {
+                    for v in acc_predicates.values() {
+                        let ae = expr_arena.get(v.node());
+                        assert!(permits_filter_pushdown(
```
Comment (nameexhaustion) [inline crates/polars-plan/src/plans/optimizer/predicate_pushdown/mod.rs:663]: drive-by - predicates that hit here should already satisfy these requirements - otherwise they should have been caught at the `Filter` node

File: crates/polars-plan/src/plans/aexpr/utils.rs [hunk 6]
```diff
@@ -1,121 +1,147 @@
...
-            function: FunctionExpr::SetSortedFlag(_),
-            ..
-        } => true,
-        AExpr::Function { options, .. } | AExpr::AnonymousFunction { options, .. } => {
-            match ctx.context {
-                Context::Default => matches!(
-                    options.collect_groups,
-                    ApplyOptions::ElementWise | ApplyOptions::ApplyList
-                ),
-                Context::Aggregation => matches!(options.collect_groups, ApplyOptions::ElementWise),
```
Comment (nameexhaustion) [inline crates/polars-plan/src/plans/aexpr/utils.rs:66]: This function was never called with `Context::Aggregation`

File: crates/polars-plan/src/plans/options.rs [hunk 7]
```diff
@@ -217,10 +209,12 @@ impl FunctionOptions {
     }
 
     pub fn is_elementwise(&self) -> bool {
-        self.collect_groups == ApplyOptions::ElementWise
-            && !self
-                .flags
-                .contains(FunctionFlags::CHANGES_LENGTH | FunctionFlags::RETURNS_SCALAR)
+        matches!(
+            self.collect_groups,
+            ApplyOptions::ElementWise | ApplyOptions::ApplyList
```
Comment (nameexhaustion) [inline crates/polars-plan/src/plans/options.rs:214]: I don't fully understand this but we apparently consider `ApplyList` as elementwise according to the test suite

File: crates/polars-plan/src/plans/aexpr/mod.rs [hunk 8]
```diff
@@ -218,35 +219,41 @@ impl AExpr {
...
+            },
+
+            Function { options, .. } => options.is_elementwise(),
+
+            Literal(v) => v.projects_as_scalar(),
+
+            Alias(_, _) | BinaryExpr { .. } | Column(_) | Ternary { .. } | Cast { .. } => true,
+
+            Agg { .. }
+            | Explode(_)
```
Comment (ritchie46) [inline crates/polars-plan/src/plans/aexpr/mod.rs:249]: Filter, gather and explode don't seem elementwise.

File: crates/polars-plan/src/plans/aexpr/utils.rs [hunk 9]
```diff
@@ -1,121 +1,147 @@
...
+            function: FunctionExpr::ArrayExpr(ArrayFunction::Get(false)),
+            ..
+        } => false,
+        // TODO: There are a lot more functions that should be caught here.
+        ae => is_elementwise(stack, ae, expr_arena),
+    }
+}
+
+pub fn permits_filter_pushdown_rec<'a>(mut ae: &'a AExpr, expr_arena: &'a Arena<AExpr>) -> bool {
+    let mut stack = vec![];
```
Comment (ritchie46) [inline crates/polars-plan/src/plans/aexpr/utils.rs:132]: can be unitvec

File: crates/polars-plan/src/plans/aexpr/utils.rs [hunk 10]
```diff
@@ -1,121 +1,147 @@
...
 }
 
-impl IsStreamableContext {
-    pub fn new(ctx: Context) -> Self {
-        Self {
-            flags: StreamableFlags::all(),
-            context: ctx,
+/// Recursive variant of `is_elementwise`
+pub fn is_elementwise_rec<'a>(mut ae: &'a AExpr, expr_arena: &'a Arena<AExpr>) -> bool {
+    let mut stack = vec![];
```
Comment (ritchie46) [inline crates/polars-plan/src/plans/aexpr/utils.rs:41]: can be unitvec

File: crates/polars-plan/src/plans/aexpr/utils.rs [hunk 11]
```diff
@@ -1,121 +1,147 @@
...
-            predicate,
-        } => {
-            !has_aexpr(*truthy, expr_arena, has_series_or_range)
-                && !has_aexpr(*falsy, expr_arena, has_series_or_range)
-                && !has_aexpr(*predicate, expr_arena, has_series_or_range)
-        },
+/// Recursive variant of `is_elementwise` that also forbids casting to categoricals. This function
+/// is used to determine if an expression evaluation can be vertically parallelized.
+pub fn is_elementwise_rec_no_cat_cast<'a>(mut ae: &'a AExpr, expr_arena: &'a Arena<AExpr>) -> bool {
+    let mut stack = vec![];
```
Comment (ritchie46) [inline crates/polars-plan/src/plans/aexpr/utils.rs:61]: can be unitvec

File: crates/polars-plan/src/plans/options.rs (hunk 7 above)
Comment (ritchie46) [reply crates/polars-plan/src/plans/options.rs:214]: I think we should see when we hit this. I believe only in a few python udf cases. I think we would never hit the streaming engine with this.

Comment (nameexhaustion) [reply crates/polars-plan/src/plans/options.rs:214]: Yes, this is hit by `map_elements()` / `map_batches(.., agg_list=True)`. This does hit the streaming engines because `FunctionOptions::is_elementwise()` gets used by all of the engines during physical plan creation / IR lowering (indirectly through `fn is_elementwise(_rec)`).

I don't think `ApplyList` is strictly elementwise, but both the in-memory and existing streaming engine currently expect it to be identified as such, so I've made it the default here. For the new-streaming engine I added an override below.
